### PR TITLE
[FLINK-4711] Let the Task trigger partition state requests and handle their responses

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR makes changes the partition state check in a way that the Task is now responsible
for triggering the state check instead of the SingleInputGate. Furthermore, the operation
returns a future containing the JobManager's answer. That way we don't have to route the
response through the TaskManager and can add automatic retries in case of a timeout.

The PR removes the JobManagerCommunicationFactory and gets rid of the excessive
PartitionStateChecker and ResultPartitionConsumableNotifier creation. Instead of creating
for each SingleInputGate one PartitionStateChecker we create one for the TaskManager which
is reused across all SingleInputGates. The same applies to the
ResultPartitionConsumableNotifier.

This PR is also a simplification for the Flip-6 implementation.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/JobManagerCommunicationFactory.java [hunk 1]
```diff
@@ -27,15 +26,6 @@
 public interface JobManagerCommunicationFactory {
```
Comment (uce) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/JobManagerCommunicationFactory.java]: This has been become obsolete with your change and is unused. Let's remove it completely?

File: flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskControlMessages.scala [hunk 2]
```diff
@@ -92,16 +92,6 @@ object TaskMessages {
   // --------------------------------------------------------------------------
 
   /**
-   * Answer to a [[RequestPartitionState]] with the state of the respective partition.
-   */
-  case class PartitionState(
```
Comment (uce) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskControlMessages.scala]: Has some unused imports after removal that we could remove before merging

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala [hunk 3]
```diff
@@ -873,8 +873,7 @@ class JobManager(
       }
 
       sender ! decorateMessage(
-        PartitionState(
-          taskExecutionId,
+        new org.apache.flink.runtime.io.network.PartitionState(
```
Comment (uce) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: I think it's not ambiguous anymore after removal of the TaskControlMessage and we can directly use `PartitionState`?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/JobManagerCommunicationFactory.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/JobManagerCommunicationFactory.java]: Good catch. Will remove it.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskControlMessages.scala (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/messages/TaskControlMessages.scala]: Yes that is also true. Will fix it.

File: flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala (hunk 3 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/scala/org/apache/flink/runtime/jobmanager/JobManager.scala]: Yes that's true. Will fix it.

**Comment by tillrohrmann** [comment]
Thanks for the review @uce. If Travis gives green light, I'll merge the PR.

**Comment by tillrohrmann** [comment]
Travis passed locally. Will merge the PR.

**Pull request closed without merging**